src/seccomp.c

```
@@ -22,24 +22,317 @@
+
+// needed for search mappings
 #include <sys/prctl.h>
+#include <sys/resource.h>
+#include <sys/socket.h>
+#include <sys/types.h>
 
 #include <seccomp.h>
 
 #include "utils.h"
 
+// libseccomp maximum per ARG_COUNT_MAX in src/arch.h
+#define SC_ARGS_MAXLENGTH	6
 #define SC_MAX_LINE_LENGTH	82	// 80 + '\n' + '\0'
 
 char *filter_profile_dir = "/var/lib/snapd/seccomp/profiles/";
+struct hsearch_data map_htab;
+
+enum parse_ret {
+	PARSE_INVALID_SYSCALL = -2,
+	PARSE_ERROR = -1,
+	PARSE_OK = 0,
+};
 
 struct preprocess {
 	bool unrestricted;
 	bool complain;
 };
 
+/*
+ * arg_cmp contains items of type scmp_arg_cmp (from SCMP_CMP macro) and
+ * length is the number of items in arg_cmp that are active such that if
+ * length is '3' arg_cmp[0], arg_cmp[1] and arg_cmp[2] are used, when length
+ * is '1' only arg_cmp[0] and when length is '0', none are used.
+ */
+struct seccomp_args {
+	int syscall_nr;
+	unsigned int length;
+	struct scmp_arg_cmp arg_cmp[SC_ARGS_MAXLENGTH];
+};
+
+/*
+ * Setup an hsearch map to map strings in the policy (eg, AF_UNIX) to
+ * scmp_datum_t values. Abstract away hsearch implementation behind map_*
+ * functions in case we want to swap this out.
+ *
+ * map_init()		- initialize the hash map
+ * map_add(key, value)	- add key/value pair to the map. Value is scmp_datum_t
+ * map_search(s)	- if found, return scmp_datum_t for key, else set errno
+ * map_destroy()	- destroy the hash map
+ */
+scmp_datum_t map_search(char *s)
+{
+	ENTRY e;
+	ENTRY *ep = NULL;
+	scmp_datum_t val = 0;
+	errno = 0;
+
+	e.key = s;
+	if (hsearch_r(e, FIND, &ep, &map_htab) == 0)
+		die("hsearch_r failed");
+
+	if (ep != NULL)
+		val = (scmp_datum_t) (ep->data);
+	else
+		errno = EINVAL;
+
+	return val;
+}
+
+void map_add(char *key, void *data)
+{
+	ENTRY e;
+	ENTRY *ep = NULL;
+	errno = 0;
+
+	e.key = key;
+	e.data = data;
+	if (hsearch_r(e, ENTER, &ep, &map_htab) == 0)
+		die("hsearch_r failed");
+
+	if (ep == NULL)
+		die("could not initialize map");
+}
```

If we make a typo in the list of calls to map_add() and call it twice with the same key value, hsearch_r() will silently ignore the ENTER operation. Lets be a little proactive here and make sure that doesn't happen. We'll need to die() if ep->data does not equal e.data.